### PR TITLE
some patch code for tab and table

### PR DESCRIPTION
**Pull request by ddkwork**

### CODE REVIEW

File: panel.go [hunk 1]
```diff
@@ -68,6 +69,10 @@ type Panel struct {
 	TooltipImmediate     bool
 }
 
+func (p *Panel) SetParent(panel *Panel) {
```
Comment (richardwilkes) [inline panel.go:72]: As I think I've stated before, I will not add a `.SetParent(*Panel)` method. If you want to set the parent, call `.AddChild(Paneler)` or `.AddChildAtIndex(Paneler,int)` on the parent instead. This will ensure all of the other things that clients expect to happen, do happen, unlike with what this method is doing.

File: panel.go [hunk 2]
```diff
@@ -15,6 +15,7 @@ import (
 	"strings"
 
 	"github.com/richardwilkes/toolbox"
+
```
Comment (richardwilkes) [inline panel.go:18]: Remove this empty line.

File: dock_header.go [hunk 3]
```diff
@@ -13,6 +13,7 @@ import (
 	"strconv"
 
 	"github.com/richardwilkes/toolbox/i18n"
+
```
Comment (richardwilkes) [inline dock_header.go:16]: Remove this empty line.

File: dock_header.go [hunk 4]
```diff
@@ -49,6 +50,11 @@ type dockHeader struct {
 	dragInsertIndex       int
 }
 
+func (d *dockHeader) CurrentDockTabLabel() *Label {
```
Comment (richardwilkes) [inline dock_header.go:53]: Why is this needed? If you want to know the title of the current tab, you can call `.CurrentDockable().Title()`. There is no need to access it from here.

File: dock_container.go [hunk 5]
```diff
@@ -50,6 +50,9 @@ func NewDockContainer(dock *Dock, dockable Dockable) *DockContainer {
 	d.AddChild(d.content)
 	return d
 }
+func (d *DockContainer) Header() *dockHeader {
```
Comment (richardwilkes) [inline dock_container.go:53]: Returning a private struct from a public API is not acceptable. Why do you feel you need to get access to this private, internal, field? It is being protected specifically to help prevent unexpected changes to its state and as such, I'm unlikely to want it exposed.

**Review by richardwilkes — COMMENTED** [review]
Currently, I'm not seeing anything in this PR that I'd want to accept. Happy to discuss why you think you need these additions, though.

File: dock_container.go (hunk 5 above)
Comment (ddkwork) [reply dock_container.go:53]: Without this action there is no way to fetch the dock title's panel

**Comment by ddkwork** [comment]
The reason for implementing the ability to close all tabs with one click is because I often have a lot of tabs open, like looking at gcs engineering, and while I don't understand what game that is, it's too much of a hassle to close them one by one when I have a lot of forms open within the gcs interface. In addition, code display widgets or edit widgets may also open a lot of tabs, if the function is improved later, it is possible to open a lot of tabs like using goland. So the context menu to close all tabs with one click is very useful for me. Anyway, I tested it on my side and it works fine, so I will close this pr if you think it's not working.

**Comment by richardwilkes** [comment]
While I won't take the changes in this PR, I did add a contextual menu into the dock tabs which has two options when more than one tab is present in a dock container:

- Close Other Tabs
- Close All Tabs

Note that these options only apply to the current dock container and not to any others that may be present elsewhere in the dock.
